hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go

```diff
@@ -0,0 +1,565 @@
+	}
+
+	return result
+}
+
+func (builder *convertFromArmBuilder) functionDeclaration() *ast.FuncDecl {
+
+	return astbuilder.DefineFunc(
+		astbuilder.FuncDetails{
+			Name:          ast.NewIdent(builder.methodName),
+			ReceiverIdent: builder.receiverIdent,
+			ReceiverType: &ast.StarExpr{
+				X: builder.receiverTypeExpr,
+			},
+			Comment: "converts from an Azure ARM object to a Kubernetes CRD object",
+			Params: []*ast.Field{
+				{
+					Type: &ast.SelectorExpr{
+						X:   ast.NewIdent(astmodel.GenRuntimePackageName),
+						Sel: ast.NewIdent("KnownResourceReference"),
+					},
+					Names: []*ast.Ident{
+						ast.NewIdent("owner"),
+					},
+				},
+				{
+					Type: ast.NewIdent("interface{}"),
+					Names: []*ast.Ident{
+						builder.inputIdent,
+					},
+				},
+			},
+			Returns: []*ast.Field{
+				{
+					Type: ast.NewIdent("error"),
+				},
+			},
+			Body: builder.functionBodyStatements(),
+		})
+}
+
+func (builder *convertFromArmBuilder) functionBodyStatements() []ast.Stmt {
+	var result []ast.Stmt
+
+	// perform a type assert and check its results
+	result = append(result, builder.assertInputTypeIsArm()...)
+
+	// Do all of the assignments for each property
+	result = append(
+		result,
+		generateTypeConversionAssignments(
+			builder.armType,
+			builder.kubeType,
+			builder.propertyConversionHandler)...)
+
+	// Return nil error if we make it to the end
+	result = append(
+		result,
+		&ast.ReturnStmt{
+			Results: []ast.Expr{
+				ast.NewIdent("nil"),
+			},
+		})
+
+	return result
+}
+
+func (builder *convertFromArmBuilder) assertInputTypeIsArm() []ast.Stmt {
+	var result []ast.Stmt
+
+	// perform a type assert
+	result = append(
+		result,
+		astbuilder.TypeAssert(builder.typedInputIdent, builder.inputIdent, builder.armTypeIdent))
+
+	// Check the result of the type assert
+	result = append(
+		result,
+		astbuilder.ReturnIfNotOk(
+			astbuilder.FormatError(
+				fmt.Sprintf("\"unexpected type supplied for FromArm function. Expected %s, got %%T\"", builder.armTypeIdent.Name),
+				builder.inputIdent)))
+
+	return result
+}
+
+//////////////////////
+// Conversion handlers
+//////////////////////
+
+func (builder *convertFromArmBuilder) namePropertyHandler() propertyConversionHandler {
+
+	return propertyConversionHandler{
+		finder: func(toProp *astmodel.PropertyDefinition, fromType *astmodel.ObjectType) (bool, *astmodel.PropertyDefinition) {
+			if toProp.Equals(GetAzureNameProperty(builder.idFactory)) && builder.isResource {
+				// Check to make sure that the ARM object has a "Name" property (which matches our "AzureName")
+				fromProperty, ok := fromType.Property(astmodel.PropertyName("Name"))
+				if !ok {
+					panic("Arm resource missing property Name")
+				}
+				return true, fromProperty
+			}
+
+			return false, nil
+		},
+		matchHandler: func(toProp *astmodel.PropertyDefinition, fromProp *astmodel.PropertyDefinition) []ast.Stmt {
+			result := astbuilder.SimpleAssignment(
```

Comment:
Yes I thought this as well. It's somewhat difficult because everything is AST types and so it's hard to just do templates in one place (you'd need to then AST parse them so you have the expressions/etc and could pass them up).

Some stuff like looping/etc is difficult in templates too (although some template languages do support that sort of thing).